### PR TITLE
[Enhancement #52] Support for debug mode

### PR DESCRIPTION
**Pull request by LaChope**

### CODE REVIEW

File: src/components/Question.jsx [hunk 1]
```diff
@@ -207,7 +207,7 @@ export default class Question extends React.Component {
 
     if (this.context.options.debugMode && !FormUtils.hasAnswer(question)) {
       return (
-          <Accordion activeKey={this.state.expanded ? question['@id'] : undefined} className="answerable-section">
+          <Accordion activeKey={this.state.expanded ? question['@id'] : this.setState({expanded: true})} className="answerable-section">
```
Comment (blcham) [inline src/components/Question.jsx:210]: hvery strange, what does this code actually mean ? Very strange place to run setState(). Will be called on each render() method call ...

So why not only having <Accordion activeKey={question['@id']} -- how is solution that you proposed better ?

Comment (LaChope) [reply src/components/Question.jsx:210]: Yes indeed, I just did not think about it

File: src/components/Question.jsx [hunk 2]
```diff
@@ -107,28 +120,42 @@ export default class Question extends React.Component {
...
+
+  renderQuestion(question) {
     if (FormUtils.isAnswerable(question) && !FormUtils.isSection(question)) {
       if (PRETTY_ANSWERABLE_LAYOUT) {
         return (
-          <div id={question['@id']}>
-            <div className="panel-title answerable-question">{this.renderAnswers()}</div>
-            <div className="answerable-subquestions">{this.renderSubQuestions()}</div>
-          </div>
+            <div id={question['@id']}>
```
Comment (blcham) [inline src/components/Question.jsx:148]: uff there are a lot of formating here ... IMO, next time its good practice not to commit it together (hope it won't be too messy merging it with pretifier

File: src/components/Question.jsx [hunk 3]
```diff
@@ -107,28 +120,42 @@ export default class Question extends React.Component {
...
     const question = this.props.question;
+    const subQuestion = question[Constants.HAS_SUBQUESTION];
     const options = this.context.options;
-
+    const renderQuestion = this.renderQuestion(question);
     if (FormUtils.isHidden(question)) {
       return null;
     }
+    if (!FormUtils.isRelevant(question) &&
+        (this.context.options.debugMode || FormUtils.isQuestionMatchingStartingQuestionId(subQuestion, options.startingQuestionId))) {
```
[the remark this answers is not in the document]
Comment (blcham) [reply src/components/Question.jsx:130]: Or put it as static of Question with relevant parameters if you need to use it in wizard step. Or use FormUtils with proper params.

File: src/components/Question.jsx [hunk 4]
```diff
@@ -463,22 +505,26 @@ export default class Question extends React.Component {
...
-      let element = React.createElement(component, {
-        key: 'sub-question-' + i,
-        question: question,
-        onChange: this.onSubQuestionChange,
-        index: i
-      });
-
+      let element = null;
+
+      if (debugMode || classname !== "show-irrelevant" || (!debugMode && questionId === startingQuestionId)) {
```
Comment (blcham) [inline src/components/Question.jsx:520]: `questionId === startingQuestionId`

IMHO, Above we implemented this functionality differently, don't see reason why it should be different.

**Review by blcham — CHANGES_REQUESTED** [review]
Having some comments to improve understandability/maintenance of the code.

File: src/components/Question.jsx (hunk 2 above)
Comment (LaChope) [reply src/components/Question.jsx:148]: Agree, hopefully Prettier will help with this

File: src/components/Question.jsx [hunk 5]
```diff
@@ -107,28 +120,42 @@ export default class Question extends React.Component {
 
   render() {
     const question = this.props.question;
+    const subQuestion = question[Constants.HAS_SUBQUESTION];
     const options = this.context.options;
-
+    const renderQuestion = this.questionComponent(question);
```
Comment (blcham) [inline src/components/Question.jsx:125]: noooooooooo :) 

opposite way around :). You have two things in this code:
1) object called questionComponent
2) function called renderQuestion with parameter question

And you just call the function to retrieve the object:
```
const questionComponent = this.renderQuestion(question);
```

I.e.:
-  functions always do: "render question", "get question", "save question"
- objects always are something: "question component", "rendered question", "question", "saved question"

File: test/rendering/TestApp.jsx [hunk 6]
```diff
@@ -63,6 +63,7 @@ class TestApp extends React.Component {
       enableForwardSkip: true,
       ...getP('startingQuestionId', 'layout-options-65'),
       startingStep: 1,
+        debugMode: true,
```
Comment (blcham) [inline test/rendering/TestApp.jsx:66]: We want to have typical usage of the form in TestApp and thus we want:
```suggestion
        debugMode: false,
```

**Review by blcham — CHANGES_REQUESTED** [review]
LGTM if fixed mentioned comments

**Pull request merged**